features/dd-sdk-android-session-replay/src/main/kotlin/com/datadog/android/sessionreplay/internal/recorder/webview/WebViewBrowserSnapshotHandler.kt

```diff
@@ -0,0 +1,59 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2016-Present Datadog, Inc.
+ */
+
+package com.datadog.android.sessionreplay.internal.recorder.webview
+
+import android.webkit.WebView
+import com.datadog.android.sessionreplay.internal.utils.RumContextProvider
+import java.util.concurrent.TimeUnit
+
+internal class WebViewBrowserSnapshotHandler(
+    private val rumContextProvider: RumContextProvider,
+    private val dataPurgeTtlLimitInNs: Long = DATA_PURGE_TTL_LIMIT_IN_NS
+) {
+
+    internal val webViewsFullSnapshotState: MutableMap<Int, Pair<String, Long>> = mutableMapOf()
+
+    internal fun triggerFullSnapshotIfNeeded(webView: WebView) {
+        val webViewId = System.identityHashCode(webView)
+        val currentRumViewId = rumContextProvider.getRumContext().viewId
+        val fullSnapshotInfo = webViewsFullSnapshotState[webViewId]
+            ?: Pair(null, System.nanoTime())
+        if (fullSnapshotInfo.first != currentRumViewId) {
+            webViewsFullSnapshotState[webViewId] = Pair(currentRumViewId, System.nanoTime())
+            webView.evaluateJavascript(FORCE_SNAPSHOT_JS_SIGNATURE, null)
+        }
+        purgeWebViewFullSnapshotStateMap()
+    }
+
+    private fun purgeWebViewFullSnapshotStateMap() {
+        val iterator = webViewsFullSnapshotState.entries.iterator()
+        @Suppress("UnsafeThirdPartyFunctionCall")
+        // next/remove can't fail: we checked hasNext and there is not concurrency access
+        // as method is called from the main thread only
+        while (iterator.hasNext()) {
+            val entry = iterator.next()
+            val timestamp = entry.value.second
+            val timeSinceLastSnapshot = System.nanoTime() - timestamp
+            if (timeSinceLastSnapshot > dataPurgeTtlLimitInNs) {
+                iterator.remove()
+            }
+        }
+        if (webViewsFullSnapshotState.size > DATA_CACHE_ENTRIES_LIMIT) {
+            webViewsFullSnapshotState.entries
+                .sortedBy { it.value.second }
+                .take(webViewsFullSnapshotState.size - DATA_CACHE_ENTRIES_LIMIT)
+                .forEach { webViewsFullSnapshotState.remove(it.key) }
+        }
+    }
+
+    companion object {
+        internal const val FORCE_SNAPSHOT_JS_SIGNATURE =
+            "window.DD_RUM.takeSessionReplayFullSnapshot();"
```

Comment:
We may have a webview/webpage without `browser-sdk` installed, in this case this call will trigger a lot of errors in the console. We first need to check that `DD_RUM` object is there, that it indeed have a session replay enabled (maybe it is now) and probably only after that call this method and maybe execute the logic in `triggerFullSnapshotIfNeeded`.

Reply:
good point...not sure how to check that though from the native side.

Reply:
We can leave native side untouched for now, and only modify this script to run `DD_RUM.takeSessionReplayFullSnapshot()` only if `DD_RUM` is defined.

Conditioning native code to run only if `DD_RUM` is defined may be a bottleneck, because even if we can get result of the script execution in the callback, the execution itself may be slow (since we need to go to JS thread, wait and come back).